File: go/0560_subarray_sum_equals_k.go
func subarraySum(nums []int, k int) int {
	cache := make(map[int]int)

	res := 0
	sum := 0
	for i := 0; i < len(nums); i++ {
		sum += nums[i]
		if sum == k {
			res += 1
		}

		if v, ok := cache[sum-k]; ok {
			res += v
		}

		if v, ok := cache[sum]; !ok {
			cache[sum] = 1
		} else {
			cache[sum] = v + 1
		}
	}
	return res
}

File: python/0560_subarray_sum_equals_k.py
from collections import defaultdict


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res = 0

        sums = [0] * len(nums)
        for i in range(len(nums)):
            sums[i] = nums[i] if i == 0 else sums[i - 1] + nums[i]
            if sums[i] == k:
                res += 1

        m = defaultdict(lambda: 0)
        for i in range(len(sums)):
            target = sums[i] - k

            res += m[target]
            m[sums[i]] += 1
        return res
